### PR TITLE
Enable torch.prod when dim is specified

### PR DESCRIPTION
**Pull request by michihirohorie**
Thanks for sending a pull request! Please make sure you read the [contributing guidelines](https://github.com/torch-spyre/torch-spyre/blob/main/CONTRIBUTING.md).

#### What type of PR is this?

- [ ] bug
- [x] feature
- [ ] documentation
- [ ] cleanup

#### What this PR does:

This PR adds support for torch.prod. Currently, it is only available when dim (and keepdim) are specified.
Calculating the product of all elements is not supported yet.

#### Which issue(s) this PR is related to:

Fixes issue #1455. Model parameter cases are addressed in this PR.

### CODE REVIEW

File: torch_spyre/_inductor/propagate_layouts.py [hunk 1]
```diff
@@ -229,19 +252,30 @@ def _single_arg_op_layout(
...
+            stick_dim += len(in_layout.size)
         # Try alternative layouts when input layout is not supported
         in_coords = host_coordinates(in_layout, dep)
         reduction_var = next(
             iter(dep.index.free_symbols - output_dep.index.free_symbols), None
         )
         layouts = []
         for in_dim in range(len(in_layout.size)):
+            if in_dim == stick_dim:
+                continue
```
Comment (yoheiueda) [inline torch_spyre/_inductor/propagate_layouts.py:255]: These changes could be more localized and less invasive. Could we localize this to the Reduction branch instead? Something like:

```python
    if isinstance(data, Reduction):
        x_dev_coords = device_coordinates(stl, dep)
        x_stick_expr = x_dev_coords[-1]
        reduction_var = next(
            iter(dep.index.free_symbols - output_dep.index.free_symbols), None
        )

        if not (
            data.reduction_type in REDUCTIONS_NON_STICK_DIM_ONLY
            and reduction_var in x_stick_expr.free_symbols
        ):
            # Try to preserve input layout
            out_stl = _output_stl_from_stick_expr(
                x_stick_expr, output, output_dep, c_size, c_stride
            )
            if out_stl is not None:
                return [out_stl]

        # Try alternative layouts when input layout is not supported
        in_coords = host_coordinates(in_layout, dep)
        out_coords = host_coordinates(output, output_dep)
        stick_dim = matching_dim(in_coords, x_stick_expr)
        layouts = []
        for in_dim in range(len(in_layout.size)):
            if in_dim == stick_dim:
                continue
            ...
```

With the constraint declared in `constants.py`:
```python
# Reduction ops that cannot reduce along the stick dimension.
REDUCTIONS_NON_STICK_DIM_ONLY = {"prod"}
```

 This keeps the common helper untouched, makes the intent clear via the constant name, and generalizes to future reductions with the same restriction.

File: torch_spyre/_inductor/lowering.py [hunk 2]
```diff
@@ -1206,3 +1206,19 @@ def inner_fn(index):
...
+
+@register_spyre_lowering(
+    torch.ops.spyre.prod_dim_int,
+    type_promotion_kind=None,
+)
+def lower_prod_dim(x, dim, keepDim=False):
+    def _prod_dim_impl(x):
+        kwargs = lowering._make_reduction_inner(
+            x, axis=[dim], keepdims=keepDim, dtype=x.dtype, override_return_dtype=None
+        )
```
Comment (yoheiueda) [inline torch_spyre/_inductor/lowering.py]:
```suggestion
def lower_prod_dim(x, dim, keepdim=False):
    def _prod_dim_impl(x):
        kwargs = lowering._make_reduction_inner(
            x, axis=[dim], keepdims=keepdim, dtype=x.dtype, override_return_dtype=None
        )
```
`keepdim` (lowercase, one word) is the PyTorch convention — matches torch.prod's public API.

File: torch_spyre/_inductor/decompositions.py [hunk 3]
```diff
@@ -904,6 +904,42 @@ def spyre_quantize_fp8_with_scale(
...
+            out_shape = out_shape[:dim] + out_shape[dim + 1 :]
+        return torch.ones(out_shape, dtype=input.dtype, device=input.device)
+
+    out_shape = out_shape[:dim] + out_shape[dim + 1 :]
+    acc = torch.ones(out_shape, dtype=input.dtype, device=input.device)
+    for i in range(size):
+        index = [slice(None)] * input.ndim
+        index[dim] = slice(i, i + 1)
+        slice_i = input[tuple(index)].squeeze(dim)
+        acc = acc * slice_i
```
Comment (yoheiueda) [inline torch_spyre/_inductor/decompositions.py]:
```suggestion
    reduce_size = out_shape.pop(dim)
    acc = torch.ones(out_shape, dtype=input.dtype, device=input.device)
    for i in range(reduce_size):
        acc = acc * input.select(dim, i)

```

Claude suggests refactoring like this. Does this make sense?

File: torch_spyre/_inductor/propagate_layouts.py (hunk 1 above)
Comment (michihirohorie) [reply torch_spyre/_inductor/propagate_layouts.py:255]: Makes sense. Thanks!

**Review by yoheiueda — APPROVED** [review]
LGTM!

File: tests/inductor/test_inductor_ops.py [hunk 4]
```diff
@@ -4285,6 +4285,66 @@ def setUp(self):
                 "3d_8x6x4": (cached_randn((2, 3, 64), dtype=torch.float16), 8, 6, 4),
             },
         },
+        # TODO: Full support for torch.prod is not yet implemented. In particular,
```
Comment (dgrove-oss) [inline tests/inductor/test_inductor_ops.py]: Is it reducing all elements that is not implemented, or is it reduction in the stick dimension that is not implemented?

File: torch_spyre/_inductor/customops.py [hunk 5]
```diff
@@ -649,3 +649,22 @@ def dequantize_fp8_with_scale(input: torch.Tensor, scale: torch.Tensor) -> torch
 def _(input: torch.Tensor, scale: torch.Tensor) -> torch.Tensor:
     # Output is FP16 with same shape as input
     return torch.empty(input.size(), dtype=torch.float16, device=input.device)
+
+
+@torch.library.custom_op("spyre::prod_dim_int", mutates_args=(), device_types="spyre")
+def prod_dim_int_fp32(
```
Comment (dgrove-oss) [inline torch_spyre/_inductor/customops.py]: Why `prod_dim_int_fp32` shouldn't it be `prod_dim_int` to match op that is being registered?

File: torch_spyre/_inductor/customops.py [hunk 6]
```diff
@@ -649,3 +649,22 @@ def dequantize_fp8_with_scale(input: torch.Tensor, scale: torch.Tensor) -> torch
...
+    input: torch.Tensor, dim: int, keepdim: bool = False
+) -> torch.Tensor:
+    pass
+
+
+@prod_dim_int_fp32.register_fake
```
Comment (dgrove-oss) [inline torch_spyre/_inductor/customops.py]: Same name question here.

**Review by dgrove-oss — COMMENTED** [review]
Overall, this PR makes sense under the assumption that we only have backend support for `prod_non_stick` and do not have support for `prod_stick` yet.

Assuming that is the case, then if we open the corresponding issue on the backend and then refer to it in the comments in this PR to document why `prod` is not following the same pattern as `sum`, etc. then I think we could go ahead and merge this.   The other goal of the comments should be to explicitly flag all the pieces of code that should be removed when we have the other ddl and can lift the restrictions.

File: tests/inductor/test_inductor_ops.py (hunk 4 above)
Comment (michihirohorie) [reply tests/inductor/test_inductor_ops.py]: Thanks for pointing that out. Clarified that the unsupported case is reduction over all elements.

File: torch_spyre/_inductor/customops.py (hunk 5 above)
Comment (michihirohorie) [reply torch_spyre/_inductor/customops.py]: Fixed, renamed it to prod_dim_int.

File: torch_spyre/_inductor/customops.py (hunk 6 above)
Comment (michihirohorie) [reply torch_spyre/_inductor/customops.py]: Also fixed.

**Comment by michihirohorie** [comment]
@dgrove-oss Thank you for your review. I opened backend issue #4409 and added references to it in the relevant code paths to document why prod currently does not follow the same pattern as other reduction ops. I also added comments to the backend-dependent restrictions introduced in this change.

**Review by dgrove-oss — APPROVED** [review]
Thanks.  lgtm.
